src/components/tl/nccl/allgatherv/allgatherv.c

```diff
@@ -0,0 +1,270 @@
+        count = ucc_coll_args_get_count(args, args->dst.info_v.counts, peer);
+        if (count != 0) {
+            displ = ucc_coll_args_get_displacement(
+                args, args->dst.info_v.displacements, peer);
+            NCCLCHECK_GOTO(ncclRecv((void *)(rbuf + displ * rdt_size),
+                                    count * rdt_size, ncclChar, peer,
+                                    team->nccl_comm, stream),
+                        exit_coll, status, UCC_TL_TEAM_LIB(team));
+        }
+    }
+    NCCLCHECK_GOTO(ncclGroupEnd(), exit_coll, status, UCC_TL_TEAM_LIB(team));
+    status = ucc_tl_nccl_collective_sync(task, stream);
+exit_coll:
+    return status;
+}
+
+ucc_status_t ucc_tl_nccl_allgatherv_p2p_init(ucc_base_coll_args_t *coll_args,
+                                             ucc_base_team_t *     team,
+                                             ucc_coll_task_t **    task_h)
+{
+    ucc_tl_nccl_team_t *nccl_team = ucc_derived_of(team, ucc_tl_nccl_team_t);
+    ucc_coll_args_t    *args      = &coll_args->args;
+    ucc_status_t        status    = UCC_OK;
+    ucc_tl_nccl_task_t *task;
+
+    CHECK_INPLACE(*args, nccl_team);
+    CHECK_USERDEFINED_DT(*args, nccl_team);
+    task = ucc_tl_nccl_init_task(coll_args, team);
+    if (!task) {
+        return UCC_ERR_NO_MESSAGE;
+    }
+    task->super.post     = ucc_tl_nccl_allgatherv_p2p_start;
+    *task_h = &task->super;
+out:
+    return status;
+}
+
+ucc_status_t ucc_tl_nccl_allgatherv_bcopy_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_nccl_task_t *task   = ucc_derived_of(coll_task, ucc_tl_nccl_task_t);
+    ucc_coll_args_t    *args   = &TASK_ARGS(task);
+    ucc_tl_nccl_team_t *team   = TASK_TEAM(task);
+    ucc_rank_t          size   = UCC_TL_TEAM_SIZE(team);
+    ucc_ee_h            ee     = coll_task->ee;
+    cudaStream_t        stream = (ee) ? (cudaStream_t) ee->ee_context :
+                                                       team->stream;
+    ucc_status_t        status = UCC_OK;
+    void               *sbuf   = args->src.info.buffer;
+    ptrdiff_t           rbuf   = (ptrdiff_t)args->dst.info_v.buffer;
+    size_t              max_count, rdt_size, sdt_size, displ, scount, rcount;
+    ucc_rank_t          peer;
+
+    task->super.super.status = UCC_INPROGRESS;
+    UCC_TL_NCCL_PROFILE_REQUEST_EVENT(coll_task, "nccl_allgatherv_start", 0);
+    scount    = args->src.info.count;
+    rdt_size  = ucc_dt_size(args->dst.info_v.datatype);
+    sdt_size  = ucc_dt_size(args->src.info.datatype);
+    max_count = ucc_coll_args_get_count(args, args->dst.info_v.counts, 0);
+    for (peer = 1; peer < size; peer++) {
+        max_count = ucc_max(ucc_coll_args_get_count(args,
+                            args->dst.info_v.counts, 0), max_count);
+    }
+    if (max_count * rdt_size > scount * sdt_size) {
+        CUDACHECK_GOTO(cudaMemcpyAsync(PTR_OFFSET(task->scratch->addr,
+                                       max_count * rdt_size * size), sbuf,
```

Comment:
don't need to copy max_count right? just scount is enough

Reply:
@vspetrov it's already copy of scount, but displacement is max_count

Reply:
@bureddy i can't do self-copy to dest buffer because if dest buffer size is smaller than max count allgather will touch unallocated memory